SSL cleanups (with a few BSD odds & ends)

dougnazar opened the pull request:
Various patches cleaning up the SSL handling:

- Free SSL structure on failed handshake (memory leak).
- Ensure we can compile without any errors or deprecations.
- Generate DH params natively on 3.x if needed (auto recommended)
- Fix several warnings about using non-length versions of `strcat`, `strcpy`, `sprintf`, etc.
- Move duplicated SSL code from `nrpe` & `check_nrpe` to separate `ssl.c` file

This has been compile tested on the following configurations: 

- Gentoo Linux (system)
  - openssl-3.0.8
- Gentoo Linux (local installation) 
  - openssl-1.0.0t
  - openssl-1.0.1u
  - openssl-1.0.2u
  - openssl-1.1.0l
  - openssl-1.1.1s
  - openssl-3.0.8
  - openssl-3.1.0
- FreeBSD 13.2
  - libressl-3.6.2
  - openssl-1.1.1t
  - openssl30-3.0.8
  - openssl31-3.1.0
- OpenBSD 7.3
  - LibreSSL 3.7.2
  - openssl-1.1.1t
  - openssl-3.0.8
  - openssl-3.1.0

It's been running in production (along with my other patches) on one of my Gentoo boxes. The others have been compile (and minimally run) tested only.

dougnazar committed:
Don't compile utils.c twice, once for each program.

dougnazar committed:
Free SSL structure on failed connection attempts.

dougnazar committed:
Fix various OpenSSL version compatibilities.

Since 1.1.0 the library will auto initialize and on 3.x these functions are deprecated.
Use ERR_get_error instead of ERR_get_error_line_data since we don't use the extra options.
Detect if library supports SSL_OP_NO_TLSv1_1 before using.

dougnazar committed:
Rewrite OpenSSL detection, use auto DH setup, and properly support 3.x.

Fix OpenSSL detection (don't include prefix and libraries may end in .a)
Add rpath to linker flags if libraries aren't in default location.
Switch to using recommended auto setup of DH parameters on OpenSSL 1.1.0+.
Rewrite OpenSSL 3.0+ generation of DH parameters to use new API.
Use OpenSSL headers to detect version since may mismatch detected binary.
Move generation of DH parameters to Makefile.

dougnazar committed:
Several small misc fixes.

dougnazar committed:
Fix various warnings.

Remove uses of strcat, strcpy, sprintf, etc.
Fix warnings about unused variables.

dougnazar committed:
Move common SSL functions to separate file.

tsadpbb commented:
Bike shedding a little bit, but could you add the ssl.o and the utils.o to the gitignore? Also, I noticed ssl.o doesn't get removed from make clean.

tsadpbb commented:
With the Makefile command
```
check_nrpe: 
        $(srcdir)/check_nrpe.c utils.o $(SRC_INCLUDE)/utils.h $(CFG_INCLUDE)/common.h $(CFG_INCLUDE)/config.h $(SSL_OBJS)
	$(CC) $(CFLAGS) -o $@ $(srcdir)/check_nrpe.c utils.o $(SSL_OBJS) $(LDFLAGS) $(SOCKETLIBS) $(SNPRINTF_O) $(OTHERLIBS)
```
Do you think you would ever need to specify the directory that utils.o is in? Something like this
```
check_nrpe: $(srcdir)/check_nrpe.c $(srcdir)/utils.o $(SRC_INCLUDE)/utils.h $(CFG_INCLUDE)/common.h $(CFG_INCLUDE)/config.h $(srcdir)/$(SSL_OBJS)
	$(CC) $(CFLAGS) -o $@ $(srcdir)/check_nrpe.c $(srcdir)/utils.o $(srcdir)/$(SSL_OBJS) $(LDFLAGS) $(SOCKETLIBS) $(SNPRINTF_O) $(OTHERLIBS)
```
This applies to the nrpe command as well

tsadpbb commented:
Should the auto_dh be enabled by default or should it be disabled by default to conform with backwards compatibility?

dougnazar commented:
> Bike shedding a little bit, but could you add the ssl.o and the utils.o to the gitignore? Also, I noticed ssl.o doesn't get removed from make clean.
> 
> Edit: Could you also add generate_dh_params to the gitignore?

I did most of my testing in separate build directories so I didn't notice. I had just assumed all `*.o` files were ignored. I'll add a wild card and `generate_dh_params`.

I missed the `ssl.o` on `clean`. Easy enough to fix.

> Do you think you would ever need to specify the directory that utils.o is in? Something like this

No. `utils.o` is in the build (current) directory, we only need to refer back to the source directory for `*.c`, `*.h` or any other shipped file.

> Should the auto_dh be enabled by default or should it be disabled by default to conform with backwards compatibility?

It's a bit of an effort to remember, but I don't think I changed backwards compatibility here. It defaulted to always generating a DH key, I just switched to auto keying on setups that supported it. It should support all the same ciphers suites and connections to older versions as before.

I believe (sorry, it's been awhile) I tested various settings of `ssl_use_adh` as well as a few older versions of `nrpe` to check the legacy cases, but my testing definitely wasn't exhaustive.

tsadpbb commented on macros/ax_nagios_get_ssl:
```diff
@@ -86,6 +96,11 @@ AC_ARG_WITH([ssl-lib],
 	AS_HELP_STRING([--with-ssl-lib=DIR],[sets location of the SSL libraries]),
 	[ssl_lib_dir=$withval])
 
+AC_ARG_ENABLE([auto_dh],
+	AS_HELP_STRING([--disable-auto-dh],[disables using builtin DH parameters (if available) and generates custom parameters]),
+	auto_dh=no,
+	auto_dh=yes)
+
```
Naturally, correct me if I am wrong about this

tsadpbb replied:
Although it might be better if auto_dh is enabled by default anyways. I'll try testing how different versions play with each other.

dougnazar replied:
Right, it defaults to using auto DH keying on setups that support it, but only if `need_dh` is true. So with the same `configure` options the software should function the exact same. The source of the DH parameters (auto or self-generated) has no affect. OpenSSL recommends using auto keying.

tsadpbb commented:
I got around to testing this and everything worked for me

sebastic commented:
`#include "ssl.h"` is problematic because this conflicts with the `ssl.h` provided by `openssl`.

The Debian package build fails because of this:
```
gcc -g -O2 -Werror=implicit-function-declaration -ffile-prefix-map=/build/nagios-nrpe-4.1.2=. -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -Wdate-time -D_FORTIFY_SOURCE=2 -I/usr/include/openssl -DHAVE_CONFIG_H -I ../include -I ./../include -c -o ssl.o ./ssl.c
./utils.c: In function 'clean_environ':
./utils.c:273:17: warning: ignoring return value of 'asprintf' declared with attribute 'warn_unused_result' [-Wunused-result]
  273 |                 asprintf(&keep, "%s,NRPE_MULTILINESUPPORT,NRPE_PROGRAMVERSION", keep_env_vars);
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./utils.c:275:17: warning: ignoring return value of 'asprintf' declared with attribute 'warn_unused_result' [-Wunused-result]
  275 |                 asprintf(&keep, "NRPE_MULTILINESUPPORT,NRPE_PROGRAMVERSION");
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./ssl.c:34:31: error: unknown type name 'SslVer'
   34 | void ssl_set_protocol_version(SslVer ssl_proto_ver, unsigned long *ssl_opts)
      |                               ^~~~~~
[...]
```
This is resolved when the header is renamed to `nrpe-ssl.h`.

sebastic commented:
These changes caused another regression, builds configured with `--disable-ssl` fail.

Fixes available in: https://github.com/NagiosEnterprises/nrpe/pull/290

dougnazar commented:
> `#include "ssl.h"` is problematic because this conflicts with the `ssl.h` provided by `openssl`.
> ```
> gcc -g -O2 -Werror=implicit-function-declaration -ffile-prefix-map=/build/nagios-nrpe-4.1.2=. -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -Wdate-time -D_FORTIFY_SOURCE=2 -I/usr/include/openssl -DHAVE_CONFIG_H -I ../include -I ./../include -c -o ssl.o ./ssl.c
> ```

It looks like specifying only `--with-ssl-lib` disabled the use of `pkg-config` and uses an older search algorithm that ends up erroneously picking `/usr/include/openssl` with no prefix, instead of using `/usr/include` and `#include <openssl/ssl.h>`.
 
> This is resolved when the header is renamed to `nrpe-ssl.h`.

Makes sense to ensure there's never a conflict. 

I did test `--with-ssl-lib` but with `--with-ssl-include` (used to test multiple OpenSSL versions in `/opt`).
I'll look into improving the include search algorithm.

dougnazar commented:
> These changes caused another regression, builds configured with `--disable-ssl` fail.
> 
> Fixes available in: #290

I did test that configuration. Hmm...

Oh, some of those fixes were in #275 which I'd written a month before. Most of my testing was with both patch sets. Should have moved those to this patch set. Sorry.
